### ISSUE TITLE
treelite model fails to load correctly on tritonserver

### ISSUE THREAD

**Issue by jperez999** [triton-inference-server/fil_backend#259]
Using `tritonserver 2.22`, the `fil backend (22.05 release)`, `sklearn==1.1.1`, `treelite 2.3.0`. when you try the following script you get a failure trying to load the model that says
```
Unavailable: Failed to allocate CUDA memory with byte size 209714400 on GPU 0: CNMEM_STATUS_OUT_OF_MEMORY 
```
However all the memory is free, verified through nvidia-smi.
The script:
```
from pathlib import Path
import treelite.sklearn
import sklearn.datasets
import sklearn.ensemble
import os

# Change to export path for model repository
export_dir = str()

# np.random.seed(123456)  # succeeds
np.random.seed(0)  # fails

rows = 200
num_features = 16
X, y = sklearn.datasets.make_regression(
    n_samples=rows,
    n_features=num_features,
    n_informative=num_features // 3,
    random_state=0,
)

# Fit RF
model = sklearn.ensemble.RandomForestRegressor()
model.fit(X, y)

config = """
name: "fil"
max_batch_size: 8192
input {
name: "input__0"
data_type: TYPE_FP32
dims: 16
}
output {
name: "output__0"
data_type: TYPE_FP32
dims: 1
}
instance_group {
}
parameters {
key: "algo"
value {
    string_value: "ALGO_AUTO"
}
}
parameters {
key: "blocks_per_sm"
value {
    string_value: "0"
}
}
parameters {
key: "model_type"
value {
    string_value: "treelite_checkpoint"
}
}
parameters {
key: "output_class"
value {
    string_value: "false"
}
}
parameters {
key: "predict_proba"
value {
    string_value: "false"
}
}
parameters {
key: "storage_type"
value {
    string_value: "AUTO"
}
}
parameters {
key: "threads_per_tree"
value {
    string_value: "1"
}
}
parameters {
key: "threshold"
value {
    string_value: "0.5000"
}
}
parameters {
key: "transfer_threshold"
value {
    string_value: "0"
}
}
parameters {
key: "use_experimental_optimizations"
value {
    string_value: "false"
}
}
backend: "fil"

"""
model_dir = Path(export_dir) / "fil"
model_dir.mkdir(parents=True, exist_ok=True)

treelite_dir = model_dir / "1"
treelite_dir.mkdir(parents=True, exist_ok=True)
treelite_path = treelite_dir / "checkpoint.tl"

with open(os.path.join(model_dir, "config.pbtxt"), "w") as f:
    f.write(config)


treelite_model = treelite.sklearn.import_model(model)
treelite_model.serialize(str(treelite_path))

```

Once completed you can try to run tritonserver against the model-repository chosen in `export_dir` and you should see a failure like the one mentioned above.

**Comment by kriskorrel-cw** [comment]
Reporting that I'm having the same issue. In some circumstances I'm unable to load a ~2MB random forest on a 24GB gpu.

I have two very similar `RandomForestClassifier`s. Both with 1000 estimators. One with 13 input features, and one with 16 input features.  
I'm able to load in the model with 16 input features but not the one with 13 input features.  
When I retrain the model of 13 input features on a subset of the 13 input features it works on some subsets and doesn't work on some other subsets. (Using a fixed random seed).

However, I found that with `storage_type=DENSE` or `storage_type=DENSE8` it does seem to load properly.

Unfortunately I'm currently not able to share a reproducible example, but hopefully the one from @jperez999 should help in debugging the issue.

**Comment by wphicks** [comment]
@jperez999 Thank you for reporting this, and @kriskorrel-cw thanks for chiming in! Apologies that we have not gotten to this sooner.

@kriskorrel-cw When you say
> However, I found that with storage_type=DENSE or storage_type=DENSE8 it does seem to load properly.
I'm assuming you actually mean `storage_type=SPARSE` or `storage_type=SPARSE8`? If so, then I think I understand what's going on here. Our `AUTO` storage_type uses a heuristic that can be fooled by models in which just a few trees or branches end up much deeper than others. In such cases, it really _ought_ to select a sparse representation, but it ends up loading as a dense model instead. In such cases, explicitly telling FIL to load sparse should give you the correct behavior.

Even in cases where a model _can_ be loaded dense, the performance benefit is quite marginal relative to a sparse representation, so there is not much downside to setting the storage_type to `SPARSE`. I'd avoid `SPARSE8` in production, though, since it's considered experimental and may impact model accuracy (unlike `SPARSE`).

I'm working on a more significant change in our in-memory model representation which should correct the issue with the `AUTO` heuristic, but in the meantime, is using `SPARSE` an acceptable workaround for both of you? If you find that your model still will not load sparse, then please let me know. If it's a _really_ big model, it may be necessary to increase Triton's CUDA memory pool, as described [here](https://nbviewer.org/github/triton-inference-server/fil_backend/blob/fea-faq_nb/notebooks/faq/FAQs.ipynb#$\color{#76b900}{\text{Example-4.2:-Increasing-Triton's-device-memory-pool}}$) in our in-progress FAQ notebook, but the memory consumed should generally not be much more than the on-disk representation of the model.

Assuming this workaround works for both of you, I'll close this issue once #266 is merged so that the workaround is more easily discoverable. We'll also make sure that the underlying issue with the `AUTO` heuristic is completely resolved with the larger ongoing work on our in-memory representation.

**Comment by kriskorrel-cw** [comment]
> @jperez999 Thank you for reporting this, and @kriskorrel-cw thanks for chiming in! Apologies that we have not gotten to this sooner.
> 
> @kriskorrel-cw When you say
> 
> > However, I found that with storage_type=DENSE or storage_type=DENSE8 it does seem to load properly.
> 
> I'm assuming you actually mean `storage_type=SPARSE` or `storage_type=SPARSE8`? If so, then I think I understand what's going on here. Our `AUTO` storage_type uses a heuristic that can be fooled by models in which just a few trees or branches end up much deeper than others. In such cases, it really _ought_ to select a sparse representation, but it ends up loading as a dense model instead. In such cases, explicitly telling FIL to load sparse should give you the correct behavior.
> 
> Even in cases where a model _can_ be loaded dense, the performance benefit is quite marginal relative to a sparse representation, so there is not much downside to setting the storage_type to `SPARSE`. I'd avoid `SPARSE8` in production, though, since it's considered experimental and may impact model accuracy (unlike `SPARSE`).
> 
> I'm working on a more significant change in our in-memory model representation which should correct the issue with the `AUTO` heuristic, but in the meantime, is using `SPARSE` an acceptable workaround for both of you? If you find that your model still will not load sparse, then please let me know. If it's a _really_ big model, it may be necessary to increase Triton's CUDA memory pool, as described [here](https://nbviewer.org/github/triton-inference-server/fil_backend/blob/fea-faq_nb/notebooks/faq/FAQs.ipynb#$%5Ccolor%7B#76b900%7D%7B%5Ctext%7BExample-4.2:-Increasing-Triton's-device-memory-pool%7D%7D$) in our in-progress FAQ notebook, but the memory consumed should generally not be much more than the on-disk representation of the model.
> 
> Assuming this workaround works for both of you, I'll close this issue once #266 is merged so that the workaround is more easily discoverable. We'll also make sure that the underlying issue with the `AUTO` heuristic is completely resolved with the larger ongoing work on our in-memory representation.

Whoops, yeah I meant `SPARSE`/`SPARSE8` 😅 

Thanks for clarifying a bit! I think that using a sparse representation would be an acceptable solution for me at the moment.

**Comment by wphicks** [comment]
> I think that using a sparse representation would be an acceptable solution for me at the moment.

Wonderful! We'll get the underlying problem sorted ASAP, but I'm glad it's not a blocker for you for now.
